Public/Src/Engine/Cache/Artifacts/LocalDiskContentStore.cs

```diff
@@ -400,7 +399,8 @@ private Possible<Unit> CreateSymlinkIfNotExistsOrTargetMismatch(AbsolutePath sym
                         }
 
                         return openResult.Succeeded;
-                    });
+                    },
+                    numberOfAttempts: OperatingSystemHelper.IsUnixOS ? 1 : 3);

```

Comment:
> I think this is going to be a problem on Windows as well. We should make a fix that addresses both operating systems.

Indeed.  I'll add a new parameter to `LocalDiskContentStore.TryCreateHandleToFile` so that the callers can decide whether this should be retried or not; the default should be "no retrying" and then Iman can tell us which caller should invoke this with retrying set to true

Reply:
The retry was necessary in the scenario where we have to hash hardlinks to the same object file. One thread can try to flush the page cache, while the other can try to open a handle to that object file for hashing. Failure in hashing a file is considered internal error, and thus we need to retry.

I think the retry should only be done if there's sharingviolation or timeout. Perhaps I was generalizing the code too much.


---
In reply to: [271538549](https://github.com/Microsoft/BuildXL/pull/76#discussion_r271538549) [](ancestors = 271538549)

Reply:
changed it to retry only on sharing violation or timeout